runtime/lua/vim/treesitter/query.lua

```diff
@@ -414,6 +414,30 @@ local directive_handlers = {
       metadata[capture_id].range = range
     end
   end,
+  ['exclude_children!'] = function(match, _, _, pred, metadata)
+    local capture_id = pred[2]
+    local node = match[capture_id]
+    local start_row, start_col, end_row, end_col = node:range()
+    local ranges = {}
+    for i = 0, node:named_child_count() - 1 do
+      local child = node:named_child(i)
+      local child_start_row, child_start_col, child_end_row, child_end_col = child:range()
+      if child_start_row > start_row or child_start_col > start_col then
+        table.insert(ranges, {
+          start_row,
+          start_col,
+          child_start_row,
+          child_start_col,
+        })
+      end
+      start_row = child_end_row
+      start_col = child_end_col
+    end
+    if end_row > start_row or end_col > start_col then
+      table.insert(ranges, { start_row, start_col, end_row, end_col })
+    end
+    metadata.content = ranges
```

Comment:
Populating `metadata.content` with random data is something we want to steer away from. Ideally we should aim to define the set of metadata fields a directive can populate, then all our ts-stdlib functions can use that data reliably.

How does this directive work? It just appears to be collecting the ranges of children.

Reply:
I'm not sure either. I just took it from https://github.com/nvim-treesitter/nvim-treesitter/commit/002084b1bea4311fdac184f080206bf890937d80

Reply:
Hmm, right ok. If we're just upstreaming code then I don't want to block it on the reasons I've said.

If we can somehow add a couple of tests and better document this internally (not publicly in the docs), then that'll make it much easier to re-implement later when we try to further standardize the metadata format.

With that said, we do have a rough TS policy that we only upstream what we use. Is that right @clason?

Reply:
> With that said, we do have a rough TS policy that we only upstream what we use

Yep, part of the motivation for opening this up is to get the discussion going how & which directives should be upstreamed. And also figure out how people are supposed to deal with directives that are not in core.  

I currently have nvim-treesitter as optional dependency and if I add this directive in my own config and then `:packadd` nvim-treesitter it fails because the directive already exists. 
I cloud work around that by using a custom directive with a different name in my query, but that doesn't seem ideal either.  Not sure what the intentions are how different plugins are supposed to use them.

Reply:
We should make nvim-treesitter check for directives and predicates before it adds them.